### DISCUSSION TITLE
What features will get you to use HostedGPT daily as your go-to over ChatGPT?

### DISCUSSION THREAD

**Discussion by florianfelsing** [AllYourBot/hostedgpt#170] [General]
Inspired by a comment by @krschacht, let's drop our ideas or actual feedback we receive in here so that we have them all in one place.

**Comment by florianfelsing** [comment]
Feedback from my wife who is _already_ using HostedGPT on a daily basis by now:

- Would love to see the copy-to-clipboard feature working (I know this is all but donw)
- Would love to see the re-generate prompt feature working
- Would love to be able to upload her user avatar to personalize her experience

**Reply by keithschacht** [reply]
This is great! I love when users start complaining, that means they care. ;) And this list is definitely different than my own so it's really helpful.

For copy-to-clipboard, we'll soon have it for the overall response, but it's another bit to get it working within markdown responses. I'm curious, is it the overall copy that your wife wants or does she also use it to generate code snippets because then we'll need to get markdown rendering & the copy icon for that.

Re-generate shouldn't be too difficult. Created Issue #173

Profile pic is a good one for us to consider. Let's get a good fallback first. I just created Issue #171 and then later we can do #172.

**Reply by florianfelsing** [reply]
My wife just needs the overall copy - I don't think that she requests markdown responses from GPT.

**Reply by keithschacht** [reply]
That one is merged in now!

**Reply by keithschacht** [reply]
And I just merged in re-generate as well.

**Reply by florianfelsing** [reply]
She says thank you :)

**Comment by stevenlafl** [comment]
- Per-conversation system messages. I make heavy use of them, but I only see a global way to change it.
- GPT Model selection (not just 3 and 4, but preview versions)
- Admin-side UI that lets you manage GPT tokens per-user (or default for global).
  - This allows you to do API usage tracking, which is useful for my family, while also preventing them from seeing what the actual token is.
- Support for browserless for things like extracting data from (asking questions about) the content on websites.
- Support for web searching.
- Editable ChatGPT response messages
- Decouple "save" and "restart from here"
- Markdown "linter": e.g. always end an opening ` ``` ` block with ` ``` ` - even if ChatGPT has not finished responding.
  - This will help with formatting problems.
  - Folders for conversation organization



The reason I'm using this is that it integrates both Claude and GPT in a single UI.

**Reply by keithschacht** [reply]
@stevenlafl That's helpful to hear. Thanks for sharing.

We can definitely add the GPT Model selection soon, but just so you know I do keep updating the app so that GPT-4 always points to the latest version so it's the new beta turbo that came out a month ago or so. Here's the Issue for this, if you want to track it: https://github.com/allyourbot/hostedgpt/issues/308

I also want to add those Admin features soon. I'm now running a server that I'm letting a few friends use and it's motivated me to add some extra controls. Here's the issue for this: https://github.com/allyourbot/hostedgpt/issues/236

Can you explain what per conversation system messages would be? If you wanted to write a system message for a single conversation, that would effectively be the same as just making your first message of the conversation be that. Is it because you want to keep a really long running conversation and you're worried about the context window?

BTW, I keep the Issue list up to date if you want to see what's on deck for the near future or if you ever want to dive in and help with anything: https://github.com/allyourbot/hostedgpt/milestone/6

**Reply by stevenlafl** [reply]
That's exactly it. Yes. And GPT 4 is supposed to pay more attention to the system prompt for directions, and it does seem to "hold memory" of it for much longer in a conversation. For example, asking for help in a refactor, putting your existing code there and then asking to revise.

Also, other things to add would be:
- support for `browserless` for things like extracting data from websites.
- if you're doing search, I know there are dedicated (paid) APIs for this to get around Google's "bot detection" limitations.

Combine these two and you might be able to add something that is similar to Perplexity or Bing.

**Reply by keithschacht** [reply]
Yep, I’m definitely going to add various agent capabilities to control things like browser. Right now I’m deep in adding voice capability and plan to tackle some website stuff soon.

Maybe on conversation system prompts you could write the first message in a chat and there is a special option underneath it, maybe an up arrow, and clicking it says “Conversation’s System Message.” Would that feel natural? It could change background color or get a border to indicate it’s special status. But what happens to your existing system message? Maybe this gets appended onto the bottom of the existing system message so you don’t lose all the default instructions you’ve added in? But then you would need some interface to edit this… ideally just editing the first message would be the interface but if we were to append then this first message would get rewritten in that moment and you’d see existing system message prepend into what’s already there.

**Reply by stevenlafl** [reply]
Interesting. If you did that, you could have the option to "pin" it as well, which helps to let you change it on the fly without scrolling. You could just only have the checkbox for system message only appear to the top message, and simply let people edit it if they want changes.

I was thinking more that the system prompt is always shown - grabbing the default from settings. If you edit it, it only applies to the current conversation. That way you could just ship the app with a default message like "You are ChatGPT, a helpful AI assistant" so it doesn't look blank next to an edit button.

Speaking of which, the edit button always triggers a regenerate (and deletes messages underneath). I actually would prefer a separate button to cause chatGPT to regenerate, rather than the current workflow.

In other words, "save" and "restart from here" should be separate. Sometimes I'd like to edit what ChatGPT said so it thinks it said something it didn't, or to make minor corrections for the history of the conversation to make sense (sometimes it has formatting errors). That helps with future me to read.

We should also possibly allow organizing things into folders.

In short,
- Decouple "save" and "restart from here"
- Editable ChatGPT response messages
- Markdown "linter": e.g. always end an opening ` ``` ` block with ` ``` ` - even if ChatGPT has not finished responding.
  - This will help with formatting problems.
- Folders for conversation organization

I've been editing my original message above to reflect all the things we've talked about.

**Comment by keithschacht** [comment]
Btw, @stephan-buckmaster just started on the assistant model settings! https://github.com/allyourbot/hostedgpt/pull/334

you can keep an eye and chime in
